djautotask/api_rest.py

```diff
@@ -0,0 +1,360 @@
+            msg = 'An error occurred: {} {}'.format(response.status_code,
+                                                    error)
+        return msg
+
+    def build_api_base_url(self):
+        self.api_base_url = '{0}v{1}/{2}/'.format(
+            self.server_url,
+            settings.AUTOTASK_CREDENTIALS['rest_api_version'],
+            self.API,
+        )
+
+    def get_headers(self):
+        headers = {'Content-Type': 'application/json'}
+
+        if self.username:
+            headers['UserName'] = self.username
+        if self.password:
+            headers['Secret'] = self.password
+        if self.integration_code:
+            headers['ApiIntegrationCode'] = self.integration_code
+
+        return headers
+
+    def build_filter(self, **kwargs):
+        for condition in kwargs['conditions']:
+            condition_arr = condition.split(",")
+            filter_obj = self._build_filter_obj(*condition_arr)
+            self.filter_array.append(filter_obj)
+
+    def _build_filter_obj(self, field, value, op='eq'):
+        filter_obj = {"op": op, "field": field, "value": value}
+        return filter_obj
+
+    def fetch_resource(self, endpoint_url, payload, full_next_url=False,
+                       retry_counter=None,
+                       *args, **kwargs):
+        """
+        Issue a GET request to the specified REST endpoint.
+
+        retry_counter is a dict in the form {'count': 0} that is passed in
+        to verify the number of attempts that were made.
+        """
+        @retry(stop_max_attempt_number=self.request_settings['max_attempts'],
+               wait_exponential_multiplier=RETRY_WAIT_EXPONENTIAL_MULTAPPLIER,
+               wait_exponential_max=RETRY_WAIT_EXPONENTIAL_MAX,
+               retry_on_exception=retry_if_api_error)
+        def _fetch_resource(endpoint_url, payload, full_next_url=False,
+                            retry_counter=None, *args, **kwargs):
+            if not retry_counter:
+                retry_counter = {'count': 0}
+            retry_counter['count'] += 1
+
+            try:
+                if full_next_url:
+                    endpoint = endpoint_url
+                else:
+                    endpoint = self._endpoint(self.POST_QUERY)
+
+                logger.debug('Making POST request to {} with {}'.format(
+                    endpoint, payload))
+                response = requests.post(
+                    endpoint,
+                    payload,
+                    timeout=self.timeout,
+                    headers=self.get_headers(),
+                )
+
+            except requests.RequestException as e:
+                logger.error('Request failed: POST {} with {}: {}'.format(
+                    endpoint, payload, e))
+                raise AutotaskAPIError('{}'.format(e))
+
+            if 200 <= response.status_code < 300:
+                rslt_items = response.json()
+                return_items.append(rslt_items.get("items"))
```

Comment:
I'd prefer to move the code that decides if there is another page and request to make out of the `_fetch_resource` method. I think it would be simpler and easier to understand to have something that is more similar to the `fetch_records` method in django-connectwise. 

Another thing that concerns me with your current implementation is that it looks like it keeps appending records to the `return_items` list for as long as records are returned from the API. I suspect this could be a problem when run with a large number of records. 

The django-connectwise `fetch_records` method doesn't have this problem since it fetches one page of records from the API, then persists the records in the TopLeft DB, and then fetches another page of records. 

So more along the lines of:
```
        while True:
            logger.info(
                'Fetching {} records'.format(
                    self.model_class.__bases__[0].__name__)
            )
            page = self.get_page()
            self.persist_page(page, results)

            next_url = page.get("pageDetails").get("nextPageUrl")
            if not next_url:
                break
```
Of course the `get_page` method would have to accept next_url somehow for `fetch_resource` to use later. Let me know if you see any problems with this approach.